File: Basics/Strings/permutation.java
import java.io.*;
import java.util.*;

public class permutation {

    public static Scanner scn = new Scanner(System.in);

	public static ArrayList<String> appendCharInString(String str, char ch){
		ArrayList<String> ans = new ArrayList<>();
		for(int i = 0; i <= str.length(); i++){
		    String s = str.substring(0,i) + ch + str.substring(i);
            ans.add(s);
		}
        return ans;
	}

    public static ArrayList<String> permutation(String str){
        ArrayList<String> ans = new ArrayList<>();
        ans.add("");
        for(int i = 0; i < str.length(); i++){
            char ch = str.charAt(i);
            ArrayList<String> smallAns = new ArrayList<>();
            for(String s : ans){
                smallAns.addAll(appendCharInString(s,ch));
            }
            ans = smallAns;
        }
        return ans;
    }

	public static void main(String[] args) {		
		System.out.println(permutation("abc"));
	}

}

File: Basics/Strings/permutation.cpp
#include<iostream>
#include<vector>
using namespace std;
vector<string> appendCharInString(string str, char ch, vector<string>& ans){
	
	for(int i = 0; i <= str.length(); i++){
	    string s = str.substr(0,i-0+1) + ch + str.substr(i);
            ans.push_back(s);
	}
    return ans;
}

vector<string> permutation(string str){
    vector<string> ans;
        ans.push_back("");
        for(int i = 0; i < str.length(); i++){
            char ch = str[i];
            vector<string> smallAns;
            for(string s : ans){
                appendCharInString(s,ch,smallAns);
            }
            ans = smallAns;
        }
        return ans;
    }

	int main() {		
        string str = "abcd";
		cout<<permutation(str);
        return 0;
	}

